book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -167,6 +167,25 @@ chains, so that the map ordering is the ordering of best to worst chains.
 
 - XXX fill in details on exact types
 
+- **Chain**: `(im::OrdMap<block::Height, Arc<Block>>, HashSet<Nullifier>, HashSet<block::Hash>)`
+  - Ord impl is ordered by work
+  - push => add a block to the end of a chain, does contextual verification
+  checks, extracts info from block for extra data sets
+  - pop => remove the lowest block
+  - fork => create a new chain fork based on a given block(hash) within
+  another chain, calls push repeatedly
+- **ChainSet**: `(BTreeSet<Chain>, BTreeMap<block::Height, Arc<Block>>)`
+  - `fn finalize(&mut self) -> Arc<Block>`
+  - `fn commit_block(&mut self, block: Arc<Block>) -> Result<(), Err>`
+    - iterate through chains, if the parent of `block` is the tip of any
+    chains, try to push block onto that chain
+    - if not, iterate through chains, checking if the parent is contained in
+    that chain
+    - if so, fork the chain at the parent and try to push `block` onto that
+    fork and add newly extended chain if the push succeeds
+    - if not, queue block, prune queued blocks that are below the reorg limit
```

Comment:
Can we look up the block (and any associated data) in a block::Hash to `(Arc<Block>, HashSet<Nullifier>)` map instead?

That way, there's no iteration, and push and fork are mostly the same operation.

Reply:
Even if they were implemented the same way (cf https://github.com/ZcashFoundation/zebra/pull/902#discussion_r485784361 ) , I don't see how we'd save the iteration here.  First, `push` and `fork` have different effects, in that one of them extends a chain, while another creates a new chain, and that difference has to be handled somehow.  Second, we have to determine which chain to try to extend with the given block, so we either have to iterate over the chains or look them up some other way.